x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java

```diff
@@ -267,6 +270,23 @@ private static boolean isCreatedOnLatestVersion(IndexMetadata indexMetadata) {
         return indexVersionCreated != null && indexVersionCreated.onOrAfter(IndexVersion.current());
     }
 
+    /**
+     * Check to see if there are any file based role mappings and if they have been loaded into cluster state. If the
+     * {@link ReservedStateMetadata} file_settings namespace contains role mapping names, it means that there should be the same number of
+     * cluster state role mappings available. If they're not available yet using {@code RoleMappingMetadata.getFromClusterState()}, they
+     * have not yet been synchronized.
+     */
+    private static boolean isReservedRoleMappingsSynced(ClusterState clusterState) {
+        ReservedStateMetadata fileSettingsMetadata = clusterState.metadata().reservedStateMetadata().get(FILE_SETTINGS_METADATA_NAMESPACE);
```

Comment:
I'd move the bit where we're retrieving handler metadata to `ReservedStateMetadata`, something like:

```java
public Set<String> keys(String handlerName) {
  ReservedStateHandlerMetadata handlerMetadata = handlers.get(handlerName);
  if (handlerMetadata == null || handlerMetadata.keys().isEmpty()) {
      return Collections.emptySet();
  }

  return Collections.unmodifiableSet(handlerMetadata.keys());
}
```

This would allow us to keep some of the logic of retrieving keys encapsulated.

I'd take it a step further and remove the `get(FILE_SETTINGS_METADATA_NAMESPACE)`, and instead just assert that we have 0 or 1 namespaces. We rely on this assumption in many other places, and I think as long as we fail very openly it will get caught by CI, should we ever add another namespace. 

Basically, just want to avoid duplicating string constants across modules. So something like:

```java
Map<String, ReservedStateMetadata> reservedStateMetadataMap = clusterState.metadata().reservedStateMetadata();
assert reservedStateMetadataMap.values().size() <= 1;
for (ReservedStateMetadata reservedStateMetadata : reservedStateMetadataMap.values()) {
    var keys = reservedStateMetadata.keys(ReservedRoleMappingAction.NAME);
    if (keys.isEmpty()) {
        var other = RoleMappingMetadata.getFromClusterState(clusterState).getRoleMappings().size();
        return other == keys.size();
    }
}
return true;
```

Reply:
I added your suggested `keys` method. That made it a lot cleaner! Thank you! 

I've update the code to use the `ReservedRoleMappingAction.NAME` instead of the new unnecessary constant I added.  

I hear you with duplicating strings, but I find the assumption and loop confusing. I think I rather keep the duplicated string for readability, but I don't feel super strongly.

Reply:
👍 if we duplicate the string, could you still add an assertion that ensures that if we get back nothing for the string, we don't have _any_ namespaces. I want to avoid us adding a namespace in the future/changing the original string/other unrelated changes to go unnoticed in the migration code

Reply:
Yes! Good idea.